File: app/src/main/java/com/apnagodam/staff/activity/SplashActivity.java
package com.apnagodam.staff.activity;

import android.Manifest;
import android.app.Dialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.apnagodam.staff.ApnaGodamApp;
import com.apnagodam.staff.Base.BaseActivity;
import com.apnagodam.staff.BuildConfig;
import com.apnagodam.staff.Network.NetworkCallbackWProgress;
import com.apnagodam.staff.Network.Response.VersionCodeResponse;
import com.apnagodam.staff.R;
import com.apnagodam.staff.databinding.ActivitySplashBinding;
import com.apnagodam.staff.db.SharedPreferencesRepository;
import com.apnagodam.staff.module.CommudityResponse;
import com.apnagodam.staff.module.UserDetails;
import com.apnagodam.staff.utils.Utility;


public class SplashActivity extends BaseActivity<ActivitySplashBinding> {
    String market_uri = "https://play.google.com/store/apps/details?id=com.apnagodam.staff&hl=en";
    private static final int REQUEST_CAMERA_CODE = 120;

    @Override
    protected int getLayoutResId() {
        Utility.changeLanguage(this, SharedPreferencesRepository.getDataManagerInstance().getSelectedLanguage());
        return R.layout.activity_splash;
    }

    @Override
    protected void setUp() {
        //getappVersion();
        getCommditityList();
        // nextMClass();
    }

    private void nextMClass() {
        if (requestPermissions()) {
            afterpermissionNext();
        } else {
            ActivityCompat.requestPermissions(
                    this,
                    new String[]{Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE},
                    REQUEST_CAMERA_CODE
            );
        }

    }

    private void afterpermissionNext() {
        UserDetails userDetails = SharedPreferencesRepository.getDataManagerInstance().getUser();
        boolean name = SharedPreferencesRepository.getDataManagerInstance().isUSerName();
        if (name) {
            startActivityAndClear(StaffDashBoardActivity.class);
        } else if (userDetails != null && (userDetails.getFname() != null) && (!userDetails.getFname().isEmpty())) {
            startActivityAndClear(StaffDashBoardActivity.class);
        } else {
            Intent intent = new Intent(SplashActivity.this, LoginActivity.class);
            intent.putExtra("setting", "");
            SplashActivity.this.startActivity(intent);
            SplashActivity.this.finish();
        }
    }

    private Boolean requestPermissions() {
        int cameraPermission = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA);
        int writeExternalPermission = ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE
        );
        int readExternalPermission = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE
        );
        return cameraPermission == PackageManager.PERMISSION_GRANTED && writeExternalPermission == PackageManager.PERMISSION_GRANTED
                && readExternalPermission == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case REQUEST_CAMERA_CODE:
                if (grantResults.length > 0) {
                    for (int i = 0; i < grantResults.length; i++) {
                        if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                            requestPermissions();
                            if (!ActivityCompat.shouldShowRequestPermissionRationale(this, permissions[i]))
                                return;
                        } else {
                            //  requestPermissions();
                            afterpermissionNext();
                        }
                    }
                    // open dialog
                }
        }
    }
   /* private void nextMClass() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                UserDetails userDetails = SharedPreferencesRepository.getDataManagerInstance().getUser();
                boolean name = SharedPreferencesRepository.getDataManagerInstance().isUSerName();
                if (name) {
                    startActivityAndClear(StaffDashBoardActivity.class);
                } else if (userDetails!=null&&(userDetails.getFname() != null) && (!userDetails.getFname().isEmpty())) {
                    startActivityAndClear(StaffDashBoardActivity.class);
                }  else {
                    Intent intent = new Intent(SplashActivity.this, LoginActivity.class);
                    intent.putExtra("setting", "");
                    SplashActivity.this.startActivity(intent);
                    SplashActivity.this.finish();
                }
            }
        }, 3000);
    }*/

    private void getappVersion() {
        apiService.getversionCode("Emp").enqueue(new NetworkCallbackWProgress<VersionCodeResponse>(getActivity()) {
            @Override
            protected void onSuccess(VersionCodeResponse body) {
                if (BuildConfig.APPLICATION_ID != null) {
                    if (Integer.parseInt(body.getVersion()) > (BuildConfig.VERSION_CODE)) {
                        showUpdateDialogue();
                    } else {
                        getCommditityList();
                    }
                }
            }
        });
    }

    private void showUpdateDialogue() {
        Dialog dialogue = new Dialog(this);
        dialogue.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialogue.setContentView(R.layout.dialogue_update_apk);
        dialogue.getWindow().setLayout(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.WRAP_CONTENT
        );
        dialogue.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        dialogue.setCancelable(false);
        dialogue.setCanceledOnTouchOutside(false);
        dialogue.setTitle(null);
        ApnaGodamApp.getApp().showUpdate = false;
        Button updateBtn = dialogue.findViewById(R.id.btn_update);
        updateBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(market_uri));
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent
                        .FLAG_ACTIVITY_NO_ANIMATION);
                startActivity(intent);
               /* if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
                        ContextCompat.checkSelfPermission(
                                SplashActivity.this,
                                Manifest.permission.READ_EXTERNAL_STORAGE
                        ) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(
                        SplashActivity.this,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE
                ) != PackageManager.PERMISSION_GRANTED
                ) {
                    ActivityCompat.requestPermissions(SplashActivity.this,
                            new String[]{Manifest.permission.READ_EXTERNAL_STORAGE,
                                    Manifest.permission.WRITE_EXTERNAL_STORAGE},
                            5);
                } else {
                    dialogue.dismiss();
                    ApnaGodamApp.getInstance().showUpdate = false;
                    downloadApk();
                }*/
            }
        });
        dialogue.show();
    }

    private void getCommditityList() {
        showDialog();
        apiService.getcommuydity_terminal_user_emp_listing("Emp").enqueue(new NetworkCallbackWProgress<CommudityResponse>(getActivity()) {
            @Override
            protected void onSuccess(CommudityResponse body) {
                if (BuildConfig.APPLICATION_ID != null) {
                    if (Integer.parseInt(body.getVersion()) > (BuildConfig.VERSION_CODE)) {
                        showUpdateDialogue();
                    } else {
//                for (int i = 0; i < body.getCategories().size(); i++) {
                        SharedPreferencesRepository.getDataManagerInstance().setCommdity(body.getCategories());
//                }
//                for (int i = 0; i < body.getTerminals().size(); i++) {
                      //  SharedPreferencesRepository.getDataManagerInstance().SETTerminal(body.getTerminals());
//                }
//                for (int i = 0; i < body.getEmployee().size(); i++) {
                        SharedPreferencesRepository.getDataManagerInstance().setEmployee(body.getEmployee());
//                }
//                for (int i = 0; i < body.getUsers().size(); i++) {
                      //  SharedPreferencesRepository.getDataManagerInstance().setUser(body.getUsers());
//                }
                        SharedPreferencesRepository.getDataManagerInstance().setContractor(body.getContractor_result());
                        hideDialog();
                        nextMClass();
                    }
                }
            }
        });
    }


}

File: app/src/main/java/com/apnagodam/staff/activity/SplashActivity.kt
package com.apnagodam.staff.activity


import android.app.Activity
import android.content.Intent
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import com.apnagodam.staff.Base.BaseActivity
import com.apnagodam.staff.Network.NetworkResult
import com.apnagodam.staff.Network.viewmodel.HomeViewModel
import com.apnagodam.staff.Network.viewmodel.NetworkSpeedViewModel
import com.apnagodam.staff.R
import com.apnagodam.staff.databinding.ActivitySplashBinding
import com.apnagodam.staff.db.SharedPreferencesRepository
import com.apnagodam.staff.utils.UpdateHelper
import com.apnagodam.staff.utils.Utility
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

@AndroidEntryPoint
class SplashActivity() : BaseActivity<ActivitySplashBinding?>() {

    //    private lateinit var inAppUpdate: InAppUpdate
    val homeViewModel by viewModels<HomeViewModel>()
    override fun getLayoutResId(): Int {
        return R.layout.activity_splash
    }



    val result =
        registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result: ActivityResult ->
            // handle callback
            if (result.resultCode != Activity.RESULT_OK) {
                UpdateHelper.checkForUpdate(this)
            }
        }

    override fun setUp() {
        UpdateHelper.checkForUpdate(this)
        getappVersion()
        nextMClass()


    }


    private fun setObservers() {
        homeViewModel.appVersionResponse.observe(this) {
            when (it) {
                is NetworkResult.Error -> {}
                is NetworkResult.Loading -> {}
                is NetworkResult.Success -> {
                    it.data?.let {
                        if (it.status == "2" || it.status == "3" || it.status == "0") {
                            startActivity(Intent(this, LoginActivity::class.java))
                        } else if (it.status == "1") {
                            getCommodityList()

                        }
                    }

                }
            }
        }
        homeViewModel.commoditiesReponse.observe(this) {
            when (it) {
                is NetworkResult.Error -> {
                    hideDialog()

                }

                is NetworkResult.Loading -> {
                    showDialog()
                }

                is NetworkResult.Success -> {
                    it.data?.let {
                        if (it.status == "2" || it.status == "3" || it.status == "0") {
                            startActivity(Intent(this, LoginActivity::class.java))
                        } else if (it.status == "1") {
                            SharedPreferencesRepository.getDataManagerInstance()
                                .setTerminals(it.terminals)
                            SharedPreferencesRepository.getDataManagerInstance()
                                .setCommdity(it.categories)
                            SharedPreferencesRepository.getDataManagerInstance().employee =
                                it.employee
                            SharedPreferencesRepository.getDataManagerInstance()
                                .setContractor(it.labourList)

                            afterpermissionNext()
                        }
                    }


                }
            }


        }
        homeViewModel.response.observe(this) {
            when (it) {
                is NetworkResult.Error -> {
                    Toast.makeText(this, it.message, Toast.LENGTH_SHORT).show();

                }

                is NetworkResult.Loading -> {
                }

                is NetworkResult.Success -> {
                    it.data?.let {
                        if (it.status == "2" || it.status == "3" || it.status == "0") {
                            startActivity(Intent(this, LoginActivity::class.java))
                        } else if (it.status == "1") {
                            SharedPreferencesRepository.getDataManagerInstance()
                                .saveUserPermissionData(it.userPermissionsResult)
                            startActivityAndClear(StaffDashBoardActivity::class.java)
                        }
                    }


                }

            }
        }
        homeViewModel.commoditiesReponse.observe(this) {
            when (it) {
                is NetworkResult.Error -> {

                }

                is NetworkResult.Loading -> {}
                is NetworkResult.Success -> {
                    it.data?.let {
                        if (it.status == "2" || it.status == "3" || it.status == "0") {
                            startActivity(Intent(this, LoginActivity::class.java))
                        } else if (it.status == "1") {
                            SharedPreferencesRepository.getDataManagerInstance()
                                .setCommdity(it.categories)
                            SharedPreferencesRepository.getDataManagerInstance().employee =
                                it.employee
                            SharedPreferencesRepository.getDataManagerInstance()
                                .setContractor(
                                    it
                                        .labourList
                                )
                        }
                    }

                }
            }
        }
    }

    private fun nextMClass() {
        setObservers()
//        networkSpeedViewModel.getNetworkSpeed(this)

    }


    private fun afterpermissionNext() {
        SharedPreferencesRepository.getDataManagerInstance().userPermission
        if (SharedPreferencesRepository.getDataManagerInstance().userPermission != null && SharedPreferencesRepository.getDataManagerInstance().userPermission.isNotEmpty()) {
            startActivityAndClear(StaffDashBoardActivity::class.java)
        } else {
            val userDetails = SharedPreferencesRepository.getDataManagerInstance().user
            if (userDetails != null && userDetails.fname != null && !userDetails.fname.isEmpty()) {
                homeViewModel.getPermissions(userDetails!!.role_id, userDetails!!.level_id)


            } else {
                val intent = Intent(this@SplashActivity, LoginActivity::class.java)
                intent.putExtra("setting", "")
                this@SplashActivity.startActivity(intent)
                finish()
            }
        }


    }


    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        afterpermissionNext()

    }

    /* private void nextMClass() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                UserDetails userDetails = SharedPreferencesRepository.getDataManagerInstance().getUser();
                boolean name = SharedPreferencesRepository.getDataManagerInstance().isUSerName();
                if (name) {
                    startActivityAndClear(StaffDashBoardActivity.class);
                } else if (userDetails!=null&&(userDetails.getFname() != null) && (!userDetails.getFname().isEmpty())) {
                    startActivityAndClear(StaffDashBoardActivity.class);
                }  else {
                    Intent intent = new Intent(SplashActivity.this, LoginActivity.class);
                    intent.putExtra("setting", "");
                    SplashActivity.this.startActivity(intent);
                    SplashActivity.this.finish();
                }
            }
        }, 3000);
    }*/
    private fun getappVersion() {
        homeViewModel.getAppVersion("Emp")


    }


    private fun getCommodityList() {
        homeViewModel.getCommodities("Emp")


    }


    companion object {
        private const val REQUEST_CAMERA_CODE = 120
        private const val REQUEST_UPDATE = 12345
    }
}
